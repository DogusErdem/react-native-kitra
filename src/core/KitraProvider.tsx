import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
