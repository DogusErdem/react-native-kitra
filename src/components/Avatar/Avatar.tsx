import { Image, View, Text } from 'react-native';
import type { AvatarProps, FCCWD } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';
import OcticonsIcon from '../Icons/Octicons';
import Badge from '../Badge/Badge';

const Avatar : FCCWD<AvatarProps> = (
  { typography,
    theme,
    source,
    size = 'medium',
    borderStyle,
    containerStyle,
    badgeContent,
    avatarIcon = <OcticonsIcon name="person" size={sizes[size].iconSize} color={theme?.primary} />,
    label,
    labelStyle,
    badgeStyle,
    badgePosition },
) => {
  const splitName = (string:string) => {
    const avatarText = string.trim().replace(/ {1,9}/g, ' ').split(' ').reduce((accumulator, currentValue:any) => accumulator.concat(currentValue[0]), '');
    return avatarText;
  };

  return (
    <View style={[{ width: sizes[size].width,
      height: sizes[size].height,
      borderRadius: borderStyle === 'rounded' ? 10 : 50,
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: theme?.primary15 }, containerStyle]}
    >
      {(() => {
        if (source) {
          return (
            <Image source={source} resizeMethod="auto" style={[{ width: sizes[size].width, height: sizes[size].height, borderRadius: borderStyle === 'rounded' ? 10 : 50 }]} />
          );
        }
        if (label) {
          return (
            <Text style={[{ fontSize: size === 'medium' ? typography?.heading.h2.fontSize : typography?.body.smedium.fontSize,
              lineHeight: size === 'medium' ? typography?.heading.h2.lineHeight : typography?.body.smedium.lineHeight,
              color: theme?.primary }, labelStyle]}
            >
              {splitName(label)}

            </Text>
          );
        }

        return (
          avatarIcon
        );
      })()}
      {badgePosition && (
      <View style={[
        { position: 'absolute', borderRadius: 50, padding: size === 'small' ? 2 : 4, right: size === 'small' ? -5 : (borderStyle === 'circular' ? 0 : -10) },
        badgePosition === 'bottom' ?
          { bottom: size === 'small' ? -2 : (borderStyle === 'circular' ? 0 : -10) } :
          { top: size === 'small' ? -5 : (borderStyle === 'circular' ? 0 : -10) }]}
      >
        <Badge
          typography={typography}
          theme={theme}
          size={size === 'small' ? 'small' : 'medium'}
          textStyles={badgeStyle?.textStyle}
          containerStyle={badgeStyle?.containerStyle}
          icon={badgeContent}
        />
      </View>
      )}

    </View>
  );
};

export default applyDefaults(Avatar);

export const sizes = {
  medium: {
    width: 100,
    height: 100,
    iconSize: 30,
    borderRadius: 100,
    fontSize: 28,
  },
  small: {
    width: 30,
    height: 30,
    iconSize: 10,
    borderRadius: 15,
    fontSize: 12,
  },
};
