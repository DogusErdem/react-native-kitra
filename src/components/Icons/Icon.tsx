import React from 'react';
import { getIconType } from '../../utilities';
import { applyDefaults } from '../../core/KitraProvider';
import type { FCCWD, IconComponentProps, IconType } from '../../types';

type IconProps = {
    type: IconType,
} & IconComponentProps;

const Icon:FCCWD<IconProps> = ({ testID, type, ...props }) => {
  const IconComponent = getIconType(type || 'ionicons');

  return (
    <IconComponent testID={testID} {...props} />
  );
};

export default applyDefaults(Icon);
