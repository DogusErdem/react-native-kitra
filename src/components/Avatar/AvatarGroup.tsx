import { View, Text } from 'react-native';
import React from 'react';
import useComponentTheme from '../../core/hooks/useComponentTheme';
import type { AvatarGroupProps, FCCWD } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';
import Avatar from './Avatar';

const AvatarGroup: FCCWD<AvatarGroupProps> = (
  { theme,
    typography,
    limitContainerStyle,
    avatars,
    avatarLimit = 10 },
) => {
  const { statusTheme } = useComponentTheme(theme, 'avatarGroup', 'default');
  const limit = (avatars?.length || 0) - avatarLimit;
  const avatarIcon = (
    <Text style={[{ fontSize: typography?.body.smedium.fontSize,
      lineHeight: typography?.body.smedium.lineHeight,
      fontWeight: '500' }, limitContainerStyle?.style, { color: statusTheme.morelabel }]}
    >
      +
      {limit}
    </Text>
  );
  return (
    <View style={{ flexDirection: 'row' }}>
      {avatars?.slice(0, avatarLimit)?.map((item, index): any => (
        <View key={item.label || item.avatarIcon?.toString()} style={{ marginRight: -7 }}>
          <Avatar
            theme={theme}
            typography={typography}
            key={index}
            source={item?.source}
            variant={item?.variant}
            avatarIcon={item?.avatarIcon}
            containerStyle={item?.containerStyle}
            size="small"
            label={item?.label}
          />

        </View>
      ))}
      {(avatars?.length || 0) > avatarLimit ? (
        <Avatar
          typography={typography}
          variant="circular"
          size="small"
          theme={{ default: { background: statusTheme.moreBackground, label: statusTheme.morelabel } }}
          avatarIcon={avatarIcon}
        />
      )
        : null}
    </View>
  );
};
export default applyDefaults(AvatarGroup);
