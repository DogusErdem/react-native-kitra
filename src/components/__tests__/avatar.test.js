import { render } from '@testing-library/react-native';
import { KitraProvider } from '../../core/KitraProvider';
import Avatar from '../Avatar/Avatar';
import Icon from '../Icons/Icon';
/* eslint-disable no-undef */

const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('Avatar', () => {
  it('renders default', () => {
    const tree = renderWithContext(
      <Avatar
        size="small"
        variant="rounded"
        label="Test"
        avatarIcon={<Icon type="octicon" name="person" size={10} color="white" />}
        source={{ uri: 'https://docs.teraportfoy.com/test-images/sliders/slider_1.jpg' }}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  describe('render with size', () => {
    it('render with image and medium size', () => {
      const { getByTestId } = renderWithContext(
        <Avatar
          size="medium"
          source={{ uri: 'https://docs.teraportfoy.com/test-images/sliders/slider_1.jpg' }}
        />,
      );
      const container = getByTestId('avatarContainer');
      const image = getByTestId('avatarImage');

      expect(container).toHaveStyle({ width: 100, height: 100 });
      expect(image).toHaveStyle({ width: 100, height: 100 });
    });
    it('render with image and small size', () => {
      const { getByTestId } = renderWithContext(
        <Avatar
          size="small"
          source={{ uri: 'https://docs.teraportfoy.com/test-images/sliders/slider_1.jpg' }}
        />,
      );
      const container = getByTestId('avatarContainer');
      const image = getByTestId('avatarImage');
      expect(container).toHaveStyle({ width: 30, height: 30 });
      expect(image).toHaveStyle({ width: 30, height: 30 });
    });
    it('render with label and medium size', () => {
      const { getByTestId } = renderWithContext(
        <Avatar
          size="medium"
          label="Test"
        />,
      );
      const container = getByTestId('avatarContainer');
      const label = getByTestId('avatarLabel');
      expect(container).toHaveStyle({ width: 100, height: 100 });
      expect(label).toHaveStyle({ fontSize: 28 });
    });
    it('render with label and small size', () => {
      const { getByTestId } = renderWithContext(
        <Avatar
          size="small"
          label="Test"
        />,
      );
      const container = getByTestId('avatarContainer');
      const label = getByTestId('avatarLabel');
      expect(container).toHaveStyle({ width: 30, height: 30 });
      expect(label).toHaveStyle({ fontSize: 12 });
    });
  });
  describe('render with variant props', () => {
    it('render with circular', () => {
      const { getByTestId } = renderWithContext(
        <Avatar
          variant="circular"
          source={{ uri: 'https://docs.teraportfoy.com/test-images/sliders/slider_1.jpg' }}
        />,
      );
      const container = getByTestId('avatarContainer');
      const image = getByTestId('avatarImage');
      expect(container).toHaveStyle({ borderRadius: 50 });
      expect(image).toHaveStyle({ borderRadius: 50 });
    });
    it('render with rounded', () => {
      const { getByTestId } = renderWithContext(
        <Avatar
          variant="rounded"
          source={{ uri: 'https://docs.teraportfoy.com/test-images/sliders/slider_1.jpg' }}
        />,
      );
      const container = getByTestId('avatarContainer');
      const image = getByTestId('avatarImage');
      expect(container).toHaveStyle({ borderRadius: 10 });
      expect(image).toHaveStyle({ borderRadius: 10 });
    });
  });
  describe('render with custom theme', () => {
    it('render with default custom theme label', () => {
      const { getByTestId } = renderWithContext(
        <Avatar
          theme={{ default: { background: 'blue', label: 'white' } }}
          label="TEST"
        />,
      );
      const container = getByTestId('avatarContainer');
      const label = getByTestId('avatarLabel');
      expect(container).toHaveStyle({ backgroundColor: 'blue' });
      expect(label).toHaveStyle({ color: 'white' });
    });
    it('render with default custom theme icon', () => {
      const { getByTestId } = renderWithContext(
        <Avatar
          theme={{ default: { background: 'blue', label: 'white' } }}
        />,
      );
      const container = getByTestId('avatarContainer');
      const icon = getByTestId('avatarIcon');
      expect(container).toHaveStyle({ backgroundColor: 'blue' });
      expect(icon).toHaveStyle({ color: 'white' });
    });
  });
  describe('render with custom styles', () => {
    it('render with containerStyle and labelSyle props', () => {
      const { getByTestId } = renderWithContext(
        <Avatar
          label="TEST"
          labelStyle={{ fontSize: 20, fontWeight: '600', textAlign: 'center' }}
          containerStyle={{ marginTop: 20, padding: 10 }}
        />,
      );
      const container = getByTestId('avatarContainer');
      const label = getByTestId('avatarLabel');
      expect(container).toHaveStyle({ marginTop: 20, padding: 10 });
      expect(label).toHaveStyle({ fontSize: 20, fontWeight: '600', textAlign: 'center' });
    });
  });
  // it('render with icon and medium size', () => {
  //   const { getByTestId } = renderWithContext(
  //     <Avatar
  //       size="medium"
  //     />,
  //   );
  //   const container = getByTestId('container');
  //   const icon = getByTestId('icon');
  //   expect(container).toHaveStyle({ width: 100, height: 100 });
  //   expect(icon).toHaveStyle({ size: 30 });
  // });
  // it('render with icon and small size', () => {
  //   const { getByTestId } = renderWithContext(
  //     <Avatar
  //       size="small"
  //     />,
  //   );
  //   const container = getByTestId('container');
  //   const icon = getByTestId('icon');
  //   expect(container).toHaveStyle({ width: 30, height: 30 });
  //   expect(icon).toHaveStyle({ size: 10 });
  // });
});
