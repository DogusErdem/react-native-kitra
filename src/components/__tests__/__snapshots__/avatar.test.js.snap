// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders default 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 30,
          "justifyContent": "center",
          "width": 30,
        },
        undefined,
        {
          "backgroundColor": "#E6EEFF",
        },
      ]
    }
    testID="avatarContainer"
  >
    <Image
      resizeMethod="auto"
      source={
        {
          "uri": "https://docs.teraportfoy.com/test-images/sliders/slider_1.jpg",
        }
      }
      style={
        [
          {
            "borderRadius": 10,
            "height": 30,
            "width": 30,
          },
        ]
      }
      testID="avatarImage"
    />
  </View>
</View>
`;
